src/job_token/unix.rs

```diff
@@ -0,0 +1,189 @@
+        //
+        // If we're called from `make` *without* the leading + on our rule
+        // then we'll have `MAKEFLAGS` env vars but won't actually have
+        // access to the file descriptors.
+        match (
+            is_pipe(&read),
+            is_pipe(&write),
+            get_access_mode(&read),
+            get_access_mode(&write),
+        ) {
+            (
+                Some(true),
+                Some(true),
+                Some(libc::O_RDONLY) | Some(libc::O_RDWR),
+                Some(libc::O_WRONLY) | Some(libc::O_RDWR),
+            ) => {
+                let read = read.try_clone().ok()?;
+                let write = write.try_clone().ok()?;
+
+                // Set read and write end to nonblocking
+                set_nonblocking(&read)?;
+                set_nonblocking(&write)?;
+
+                Some(Self {
+                    read,
+                    write: Some(write),
+                })
+            }
+            _ => None,
+        }
+    }
+
+    pub(super) fn try_acquire(&self) -> io::Result<Option<()>> {
+        let mut fds = [libc::pollfd {
+            fd: self.read.as_raw_fd(),
+            events: libc::POLLIN,
+            revents: 0,
+        }];
+
+        let ret = cvt(unsafe { libc::poll(fds.as_mut_ptr(), 1, 0) })?;
+        if ret == 1 {
+            let mut buf = [0];
+            match (&self.read).read(&mut buf) {
+                Ok(1) => Ok(Some(())),
+                Ok(_) => Ok(None), // 0, eof
+                Err(e)
+                    if e.kind() == io::ErrorKind::Interrupted
+                        || e.kind() == io::ErrorKind::WouldBlock =>
+                {
+                    Ok(None)
+                }
+                Err(e) => Err(e),
+            }
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub(super) fn release(&self) -> io::Result<()> {
+        // Note that the fd may be nonblocking but we're going to go ahead
+        // and assume that the writes here are always nonblocking (we can
+        // always quickly release a token).
+        //
+        // For write to block, this would mean that pipe is full.
+        // If all every release are pair with an acquire, then this cannot
+        // happen.
+        //
+        // If it does happen, it is likely a bug in the program using this
+        // crate or some other programs that use the same jobserver have a
+        // bug in their  code
+        //
+        // If that turns out to not be the case we'll get an error anyway!
+        let mut write = self.write.as_ref().unwrap_or(&self.read);
+        match write.write(&[b'+'])? {
+            1 => Ok(()),
+            _ => Err(io::Error::from(io::ErrorKind::UnexpectedEof)),
+        }
+    }
+}
+
+fn set_nonblocking(file: &File) -> Option<()> {
+    // F_SETFL can only set the O_APPEND, O_ASYNC, O_DIRECT, O_NOATIME, and
+    // O_NONBLOCK flags.
+    //
+    // For pipe, only O_NONBLOCK is meaningful, so it is ok to
```

Comment:
Is this guaranteed to be true on all unix (including future ones?)

Reply:
There's no guarantee but I think it should be ok, since I read in manpage it isn't ok to change flags other than `O_NONBLOCK`

Reply:
Besides, we really don't want anyflags other than `O_NONBLOCK` to be set.

Pipe is always fifo, so `O_APPEND` means nothing.
Pipe does not write to disk, so `O_ASYNC`, `O_NOATIME` and `O_DIRECT` is useless.

`O_ASYNC` is for fs async I/O which is usually not used, `O_DIRECT` is for direct disk I/O which is also usually not used and meanless for in-memory only pipe.